packages/gatsby/src/schema/node-model.js

```diff
@@ -307,35 +320,75 @@ class LocalNodeModel {
-        result = result[0]
-        this.trackInlineObjectsInRootNode(result)
-      } else {
-        result = null
-
-        // Couldn't find matching node.
-        //  This leads to a state where data tracking for this query gets empty.
-        //  It means we will NEVER re-run this query on any data updates
-        //  (even if a new node matching this query is added at some point).
-        //  To workaround this, we have to add a connection tracking to re-run
-        //  the query whenever any node of this type changes.
-        pageDependencies.connectionType = gqlType.name
-      }
-    } else if (result) {
-      result.forEach(node => this.trackInlineObjectsInRootNode(node))
-    }
+    entries.forEach(node => this.trackInlineObjectsInRootNode(node))
 
     if (trackInlineObjectsActivity) {
       trackInlineObjectsActivity.end()
     }
+    return { gqlType, entries, totalCount }
+  }
+
+  /**
+   * Get nodes of a type matching the specified query.
+   *
+   * Note: this method returns a slice of result when `skip` and `limit` are set.
+   *
+   * @param {Object} args
+   * @param {Object} args.query Query arguments (`filter`, `sort`, `skip`, `limit`)
+   * @param {(string|GraphQLOutputType)} args.type Type
+   * @param {PageDependencies} [pageDependencies]
+   * @returns {Promise<IQueryResult>}
+   */
+  async findAll(args, pageDependencies = {}) {
+    const { gqlType, ...result } = await this._query(args, pageDependencies)
 
     // Tracking connections by default:
-    if (!firstOnly && typeof pageDependencies.connectionType === `undefined`) {
+    if (typeof pageDependencies.connectionType === `undefined`) {
       pageDependencies.connectionType = gqlType.name
     }
-    return this.trackPageDependencies(result, pageDependencies)
+    this.trackPageDependencies(result.entries, pageDependencies)
+    return result
+  }
+
+  /**
+   * Get the first node of a type matching the specified query.
+   *
+   * @param {Object} args
+   * @param {Object} args.query Query arguments (supports: `filter`)
+   * @param {(string|GraphQLOutputType)} args.type Type
+   * @param {PageDependencies} [pageDependencies]
+   * @returns {Promise<Node | null>}
+   */
+  async findOne(args, pageDependencies = {}) {
+    const { query } = args
+    if (query.sort?.fields?.length > 0) {
+      // If we support sorting and return the first node based on sorting
+      // we'll have to always track connection not an individual node
+      reporter.warn(
+        `nodeModel.findOne() does not support sorting. Use nodeModel.findAll({ query: { limit: 1 } }) instead`
+      )
+      // TODO: throw in v4
+      // throw new Error(
+      //   `nodeModel.findOne() does not support sorting. Use nodeModel.findAll({ query: { limit: 1 } }) instead`
+      // )
+    }
+    const { gqlType, entries } = await this._query({
+      ...args,
+      query: { ...query, skip: 0, limit: 1, sort: undefined },
+    })
+    const result = Array.from(entries)
```

Comment:
What dataType is entries? Is it an iterator? Does it already only have 1 item?

If it's an iterator you could also do `entries.next()` to get the next value.

Reply:
It is iterable with a single item (because of `limit: 1`). So technically we can do `entries[Symbol.iterator]().next().value` here but I don't see much benefits given there is only one item anyway.

Reply:
Wasn't sure if we got 1 item back that's why I asked :)